### PR TITLE
Batching helpers

### PR DESCRIPTION
**Pull request by tbarbette**
3 helper element to easily build batch-compatible without downgrading performances. The explanation is actually already in the wiki : https://github.com/tbarbette/fastclick/wiki/Batching

And then PaintRR which uses one of them.

### CODE REVIEW

File: include/click/batchelement.hh [hunk 1]
```diff
@@ -131,5 +131,127 @@ class BatchElement : public Element { public:
...
+
+#if HAVE_BATCH
+    void push_batch(int port, PacketBatch* head) override final {
+        head = T::simple_action_batch(head);
+        if (head)
+            output_push_batch(port,head);
+    }
+
+    PacketBatch* pull_batch(int port, unsigned max) override final {
+        PacketBatch* head = input_pull_batch(port,max);
```
Comment (gkatsikas) [inline include/click/batchelement.hh]: assert max, unless input_pull_batch() does it

File: include/click/batchelement.hh [hunk 2]
```diff
@@ -131,5 +131,127 @@ class BatchElement : public Element { public:
...
+        if (head) {
```
Comment (gkatsikas) [inline include/click/batchelement.hh]: Be consistent in parentheses for single statement conditions. Either you always use the parentheses or you never do it.

File: include/click/batchelement.hh [hunk 3]
```diff
@@ -131,5 +131,127 @@ class BatchElement : public Element { public:
...
+
+    private:
+        SimpleBatchElement(){};
+        friend T;
+
+};
+
+
+/**
+ * Batch helper element
```
Comment (gkatsikas) [inline include/click/batchelement.hh]: you need a full stop at the end of this sentence.

File: include/click/batchelement.hh [hunk 4]
```diff
@@ -131,5 +131,127 @@ class BatchElement : public Element { public:
...
+ * extend SimpleElement<T> where T is the element itself.
+ * It also avoids the virtual call, so actually a vanilla element using
+ * this version even without batching will run faster.
+ *
+ * The inherited element cannot be extended further because of CRTP !
+ */
+template <typename T>
+class SimpleElement : public SimpleBatchElement<T> { public:
+
+    inline PacketBatch* _sm_action_batch(int port, PacketBatch* batch) {
```
Comment (gkatsikas) [inline include/click/batchelement.hh]: Also be consistent with brackets. Should you always open a bracket at the same line with the function name or always put it in the next line?

File: include/click/batchelement.hh [hunk 5]
```diff
@@ -131,5 +131,127 @@ class BatchElement : public Element { public:
...
+};
+
+
+/**
+ * Batch helper element.
+ *
+ * Allows the user to simply implement simple_action and simple_action_batch
+ *  but relying on CRTP instead of virtual, removing one virtual call.
+ *
+ * The inherited element cannot be extended further because of CRTP !
```
Comment (gkatsikas) [inline include/click/batchelement.hh]: What countermeasures do we have for this restriction?
If we convert all Click elements to this template, how can we leverage inheritance?
Maybe using template parameters, but this makes the code base more complicated.

File: elements/standard/paintrr.cc [hunk 6]
```diff
@@ -0,0 +1,49 @@
+/*
+ * paintswitch.{cc,hh} -- element routes packets to one output of several
+ * Douglas S. J. De Couto.  Based on Switch element by Eddie Kohler
+ *
+ * Copyright (c) 2002 MIT
+ * Copyright (c) 2008 Meraki, Inc.
```
Comment (gkatsikas) [inline elements/standard/paintrr.cc]: Change the authorship if this is your example.

File: include/click/batchelement.hh (hunk 4 above)
Comment (tbarbette) [reply include/click/batchelement.hh]: I generally do on the same line.

File: include/click/batchelement.hh (hunk 5 above)
Comment (tbarbette) [reply include/click/batchelement.hh]: This is a fast easy way to make an element compatible. If one wants inheritance he can copy paste the code in the CRTP. Or extend the base template to pass the inherited and the final inheritance. But in any case if someone wants to do something complex he'll implement push* and pull* himself...

File: include/click/batchelement.hh (hunk 1 above)
Comment (tbarbette) [reply include/click/batchelement.hh]: I prefer not to add an assert here. Each element should not ask for 0 packet, that's a coding error. We can't add run-time checks for every of thoses.

**Pull request merged**